x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java

```diff
@@ -0,0 +1,368 @@
+    }
+
+    private static void ensureInternalIndex(Client client, String index, XContentBuilder mappings, Settings settings, String alias) {
+        GetIndexRequest getIndexRequest = new GetIndexRequest().indices(CONNECTOR_CONCRETE_INDEX_NAME);
+        ThreadContext threadContext = client.threadPool().getThreadContext();
+        executeAsyncWithOrigin(threadContext, ENT_SEARCH_ORIGIN, getIndexRequest, new ActionListener<GetIndexResponse>() {
+            public void onResponse(GetIndexResponse getIndexResponse) {
+                logger.info("Found " + index + " index.");
+            }
+
+            public void onFailure(Exception e) {
+                final Throwable cause = ExceptionsHelper.unwrapCause(e);
+                if (cause instanceof ResourceNotFoundException) {
+                    logger.error("Index " + index + " not found.");
+                    createInternalIndex(client, index, mappings, settings, alias);
+                } else {
+                    logger.error("Error getting " + index + " index " + e.toString());
+                }
+            }
+        }, client.admin().indices()::getIndex);
+    }
+
+    private static void ensurePipeline(Client client) {
+        GetPipelineRequest getPipelineRequest = new GetPipelineRequest(ENT_SEARCH_INGESTION_PIPELINE_NAME);
+        ThreadContext threadContext = client.threadPool().getThreadContext();
+        executeAsyncWithOrigin(threadContext, ENT_SEARCH_ORIGIN, getPipelineRequest, new ActionListener<>() {
+            public void onResponse(GetPipelineResponse getPipelineResponse) {
+                if (getPipelineResponse.isFound() == false) {
+                    createPipeline(client);
+                } else {
+                    logger.info("Found " + ENT_SEARCH_INGESTION_PIPELINE_NAME + " pipeline.");
+                }
+            }
+
+            public void onFailure(Exception e) {
+                final Throwable cause = ExceptionsHelper.unwrapCause(e);
+                if (cause instanceof ResourceNotFoundException) {
+                    logger.error("Pipeline " + ENT_SEARCH_INGESTION_PIPELINE_NAME + " not found.");
+                    createPipeline(client);
+                } else {
+                    logger.error("Error getting " + ENT_SEARCH_INGESTION_PIPELINE_NAME + " index " + e.toString());
+                }
+            }
+        }, client.admin().cluster()::getPipeline);
+    }
+
+    private static Settings getConnectorsIndexSettings() {
+        return Settings.builder()
+            .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+            .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+            .put(IndexMetadata.SETTING_AUTO_EXPAND_REPLICAS, "0-3")
+            .put(IndexMetadata.SETTING_INDEX_HIDDEN, true)
+            .put("index.refresh_interval", "1s")
+            .build();
+    }
+
+    private static XContentBuilder getConnectorsIndexMappings() {
```

Comment:
I vaguely remember seeing that you could define mappings either like this, or as raw JSON. I feel like it might be easier to maintain raw JSON, since we're not doing anything fancy in this construction (no conditional properties or anything). But I don't know if that's not idiomatic. Is it something you've already thought about?

Reply:
haha I was looking at this and thought, WOW so much work for something that could be a JSON resource 🤣

Reply:
Yes, but this appears to be the idiomatic way to do mappings for ES managed/system indices.